QRCoder/QRCodeGenerator.cs

```diff
@@ -254,82 +275,181 @@ private static QRCodeData GenerateQrCode(string bitString, ECCLevel eccLevel, in
+        private static BitArray GetFormatString(ECCLevel level, int maskVersion)
         {
-            var generator = "10100110111";
-            var fStrMask = "101010000010010";
+            var fStrEcc = new BitArray(15); // Total length including space for mask version and padding
+            WriteEccLevelAndVersion();
 
-            var fStr = (level == ECCLevel.L) ? "01" : (level == ECCLevel.M) ? "00" : (level == ECCLevel.Q) ? "11" : "10";
-            fStr += DecToBin(maskVersion, 3);
-            var fStrEcc = fStr.PadRight(15, '0').TrimStart('0');
-            while (fStrEcc.Length > 10)
+            int index = 0;
+            int count = 15;
+            TrimLeadingZeros();
+            while (count > 10)
             {
-                var sb = new StringBuilder();
-                generator = generator.PadRight(fStrEcc.Length, '0');
-                for (var i = 0; i < fStrEcc.Length; i++)
-                    sb.Append((Convert.ToInt32(fStrEcc[i]) ^ Convert.ToInt32(generator[i])).ToString());
-                fStrEcc = sb.ToString().TrimStart('0');
+                for (var i = 0; i < _getFormatGenerator.Length; i++)
+                    fStrEcc[index + i] ^= _getFormatGenerator[i];
+                TrimLeadingZeros();
+            }
+            ShiftTowardsBit0(fStrEcc, index);
+            fStrEcc.Length = 10 + 5;
+            ShiftAwayFromBit0(fStrEcc, (10 - count) + 5);
+            WriteEccLevelAndVersion();
+            fStrEcc.Xor(_getFormatMask);
+            return fStrEcc;
+
+            void WriteEccLevelAndVersion()
+            {
+                switch (level)
+                {
+                    case ECCLevel.L: // 01
+                        fStrEcc[1] = true;
+                        break;
+                    case ECCLevel.H: // 10
+                        fStrEcc[0] = true;
+                        break;
+                    case ECCLevel.Q: // 11
+                        fStrEcc[0] = true;
+                        fStrEcc[1] = true;
+                        break;
+                    default: // M: 00
+                        break;
+                }
+                int indexTemp = 2;
+                DecToBin(maskVersion, 3, fStrEcc, ref indexTemp);
+            }
+
+            void TrimLeadingZeros()
+            {
+                while (!fStrEcc[index])
+                {
+                    index++;
+                    count--;
+                }
             }
-            fStrEcc = fStrEcc.PadLeft(10, '0');
-            fStr += fStrEcc;
 
-            var sbMask = new StringBuilder();
-            for (var i = 0; i < fStr.Length; i++)
-                sbMask.Append((Convert.ToInt32(fStr[i]) ^ Convert.ToInt32(fStrMask[i])).ToString());
-            return sbMask.ToString();
         }
 
-        private static string GetVersionString(int version)
+        private static void ShiftTowardsBit0(BitArray fStrEcc, int num)
         {
-            var generator = "1111100100101";
+#if NETCOREAPP
```

Comment:
Why is a distinction made here between NETCOREAPP and non-NETCOREAPP? Surely the RightShift/LeftShift methods are available in all Framework versions? https://learn.microsoft.com/de-de/dotnet/api/system.linq.expressions.expression.rightshift?view=net-8.0

Reply:
No, just .NET Standard 2.1 / .NET Core 2.0 and newer -- see BitArray.LeftShift here:

https://learn.microsoft.com/en-us/dotnet/api/system.collections.bitarray.leftshift?view=net-8.0

Reply:
By the way, `BitArray` appears to be heavily optimized within the .NET runtime (albeit for longer bit runs):

https://github.com/dotnet/runtime/blob/2ad47cd52ebe57c1b4c2e28f95cc4eed6fcb354d/src/libraries/System.Collections/src/System/Collections/BitArray.cs#L664

One nice feature is that it is resizable like a `List<T>`.  However, `BitArray` is an allocation, and it's underlying `int[]` array is an allocation also.  So after this PR, I may review again and see if `BitArray` itself can be eliminated anywhere.  For example, `GetVersionString` and `GetFormatString` always manipulate and return bit arrays that are less than 32 bits.  The math could be done in an `int`, eliminating a couple allocations.  But I'm not planning to remove every allocation necessarily.  It's more important to get the recursive math optimized.